Sources/AWSSDKSwiftCore/Message/AWSMiddleware.swift

```diff
@@ -0,0 +1,93 @@
+//
+//  AWSMiddleware.swift
+//  AWSSDKSwift
+//
+//  Created by Yuki Takei/Adam Fowler on 2019/08/17.
+//
+//
+
+/// Middleware protocol. Gives ability to process requests before they are sent to AWS and process responses before they are converted into output shapes
+public protocol AWSServiceMiddleware {
+    
+    /// Process AWSRequest before it is converted to a HTTPClient Request to be sent to AWS
+    func chain(request: AWSRequest) throws -> AWSRequest
+    
+    /// Process responseBody before it is converted to an output AWSShape
+    func chain(responseBody: Body) throws -> Body
+}
+
+/// Default versions of protocol functions
+public extension AWSServiceMiddleware {
+    func chain(request: AWSRequest) throws -> AWSRequest {
+        return request
+    }
+    func chain(responseBody: Body) throws -> Body {
+        return responseBody
+    }
+}
+
+#if DEBUG
```

Comment:
In our shared dev and qa environments, we run code in production mode. It would be cool if I could enable this somehow in those envs if I wanted to. 

Maybe this code should compile in non debug envs, then allow setting the log level in the init

```
    let logLevel : LogLevel

    init(log : @escaping (String)->() = { print($0) }, logLevel: LogLevel = .info) {
        self.log = log
        self.logLevel = logLevel
    }
    
    enum LogLevel {
      case debug
      case info
      case error
    }

    /// output request
    public func chain(request: AWSRequest) throws -> AWSRequest {
        if logLevel == .debug {
          log("Request: \(request.operation)")
          log("\(request.httpMethod) \(request.url)")
          log("Headers: " + getHeadersOutput(request.httpHeaders))
          log("Body: " + getBodyOutput(request.body))
        }
        return request
    }

    /// output response
    public func chain(responseBody: Body) throws -> Body {
        if logLevel == .debug {
            var output = "Response: "
            output += getBodyOutput(responseBody)
            log(output)
        }
        return responseBody
    }

----


    let kinesisClient = AWSClient(
        accessKeyId: "foo",
        secretAccessKey: "bar",
        region: nil,
        amzTarget: "Kinesis_20131202",
        service: "kinesis",
        serviceProtocol: ServiceProtocol(type: .json, version: ServiceProtocol.Version(major: 1, minor: 1)),
        apiVersion: "2013-12-02",
        middlewares: [AWSLoggingMiddleware(logLevel: .debug)],
        possibleErrorTypes: [KinesisErrorType.self]
    )
```

I can update this PR if you like this idea

Reply:
code can be made available in release. What would you output for logLevel info and error though? If there is no difference then the way to disable the logging would be not include the middleware.